sdk/attestation/azure-security-attestation/src/main/java/com/azure/security/attestation/AttestationAsyncClient.java

```diff
@@ -237,6 +243,24 @@
     }
 
 
+    /**
+     * Return cached attestation signers, fetching from the internet if needed.
+     * @return cached signers.
+     */
+    Mono<List<AttestationSigner>> getCachedAttestationSigners() {
+        if (this.cachedSigners.get() != null) {
+            return Mono.just(this.cachedSigners.get());
+        } else {
+            return this.signerImpl.getAsync()
```

Comment:
>getAsync [](http://example.com/codeflow?start=35&length=8)

Can this not cause the async call to be made more than once - is that intentional?

Reply:
An excellent question... A lot of this depends on the threading model for Java applications with React, I've not been able to find any good documentation about how Java's concurrency model works w.r.t. threading.

In this case, I followed the caching model I found in a number of other Azure SDKs, assuming that they did the right thing (I can chase down which SDKs I used if it helps).

Having said that, the data returned is the same for each instant, so the possible cost of a concurrency failure is to retrieve the data twice.

Note that the `cachedSigners` value is only ever updated once - if we DO have a double fetch call, then the `.compareandset` will only cause the cached value to be updated once atomically.

Reply:
Reactor is a 'reactive model' rather than a traditional async\await - so it is written such that one can never make assumption on when an info is fetched.

So, even though it supports complex scenarios like sync which can be used for true caching, what you have is the cheapest way to achieve the same. 

As you mentioned the only concern would be if two calls were a problem which it is not, so we are good here.